SofaKernel/modules/SofaEigen2Solver/src/SofaEigen2Solver/EigenBaseSparseMatrix_MT.h

```diff
@@ -0,0 +1,220 @@
+  typedef typename Lhs::Index Index;
+  static void run(const SparseLhsType& lhs, const DenseRhsType& rhs, DenseResType& res, const typename Res::Scalar& alpha, unsigned nbThreads)
+  {
+#pragma omp parallel for num_threads(nbThreads)
+    for(Index j=0; j<lhs.outerSize(); ++j)
+    {
+      typename Res::RowXpr res_j(res.row(j));
+      for(LhsInnerIterator it(lhs,j); it ;++it)
+        res_j += (alpha*it.value()) * rhs.row(it.index());
+    }
+  }
+};
+
+template<typename SparseLhsType, typename DenseRhsType, typename DenseResType>
+struct sparse_time_dense_product_impl_MT<SparseLhsType,DenseRhsType,DenseResType, ColMajor, false>
+{
+  typedef typename internal::remove_all<SparseLhsType>::type Lhs;
+  typedef typename internal::remove_all<DenseRhsType>::type Rhs;
+  typedef typename internal::remove_all<DenseResType>::type Res;
+  typedef typename Lhs::InnerIterator LhsInnerIterator;
+  typedef typename Lhs::Index Index;
+  static void run(const SparseLhsType& lhs, const DenseRhsType& rhs, DenseResType& res, const typename Res::Scalar& alpha, unsigned nbThreads)
+  {
+    #pragma omp parallel for num_threads(nbThreads)
+    for(Index j=0; j<lhs.outerSize(); ++j)
+    {
+      typename Rhs::ConstRowXpr rhs_j(rhs.row(j));
+      for(LhsInnerIterator it(lhs,j); it ;++it)
+        res.row(it.index()) += (alpha*it.value()) * rhs_j;
+    }
+  }
+};
+
+template<typename SparseLhsType, typename DenseRhsType, typename DenseResType,typename AlphaType>
+inline void sparse_time_dense_product_MT(const SparseLhsType& lhs, const DenseRhsType& rhs, DenseResType& res, const AlphaType& alpha, unsigned nbThreads)
+{
+  sparse_time_dense_product_impl_MT<SparseLhsType,DenseRhsType,DenseResType>::run(lhs, rhs, res, alpha, nbThreads);
+}
+
+} // end namespace internal
+
+template<typename Lhs, typename Rhs>
+class SparseTimeDenseProduct_MT
+  : public ProductBase<SparseTimeDenseProduct_MT<Lhs,Rhs>, Lhs, Rhs>
+{
+    unsigned m_nbThreads;
+  public:
+    EIGEN_PRODUCT_PUBLIC_INTERFACE(SparseTimeDenseProduct_MT)
+
+    SparseTimeDenseProduct_MT(const Lhs& lhs, const Rhs& rhs, unsigned nbThreads) : Base(lhs,rhs), m_nbThreads(nbThreads)
+    {}
+
+    template<typename Dest> void scaleAndAddTo(Dest& dest, const Scalar& alpha) const
+    {
+#ifdef _OPENMP
+        // no multithreading for too small vectors
+        if( ( m_rhs.cols() == 1 && m_rhs.rows()<3000 ) || m_nbThreads==1 )
+            internal::sparse_time_dense_product(m_lhs, m_rhs, dest, alpha);
+        else
+        {
+//            msg_info()<<"SparseTimeDenseProduct_MT: "<<m_nbThreads<<std::endl;
+            internal::sparse_time_dense_product_MT<Lhs,Rhs,Dest,Scalar>(m_lhs, m_rhs, dest, alpha, m_nbThreads);
+        }
+#else
+        internal::sparse_time_dense_product(m_lhs, m_rhs, dest, alpha);
+#endif
+    }
+
+  private:
+    SparseTimeDenseProduct_MT& operator=(const SparseTimeDenseProduct_MT&);
+};
+
+
+
+} // namespace Eigen
+
+
+namespace sofa
+{
+
+namespace component
+{
+
+namespace linearsolver
```

Comment:
any idea why this file is not identified as "moved/renamed" by git?

otherwise maybe put nasted namespace.

Reply:
no idea